### PR TITLE
pid1: don't return any error in manager_dispatch_notify_fd()

### PR DESCRIPTION
**Pull request by fbuihuu**
If manager_dispatch_notify_fd() fails and returns an error then the handling of
service notifications will be disabled entirely leading to a compromised system.

For example pid1 won't be able to receive the WATCHDOG messages anymore and
will kill all services supposed to send such messages.

### CODE REVIEW

File: src/core/manager.c [hunk 1]
```diff
@@ -1716,10 +1716,14 @@ static int manager_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t
 
         n = recvmsg(m->notify_fd, &msghdr, MSG_DONTWAIT|MSG_CMSG_CLOEXEC);
         if (n < 0) {
-                if (errno == EAGAIN || errno == EINTR)
-                        return 0;
+                if (errno != EAGAIN && errno != EINTR)
```
Comment (keszybz) [inline src/core/manager.c]: `!IN_SET(errno, …)`

Comment (fbuihuu) [reply src/core/manager.c]: ok I'm updating this part

**Comment by fbuihuu** [comment]
@keszybz: branch updated.

That said I'm wondering if we couldn't improve the sd_event API in order to allow event handlers to return errors (which would be logged) without being disabled ...

**Comment by keszybz** [comment]
If our sd_event_add_\* functions had flags, this would be easy. Now one'd need to add new functions... So this becomes a bit annoying to implement. In principle you could always do everything in the handler and then return >=0, but I think it'd often be convenient to keep the handler simple and let sd-event log the error and continue. Especially that sd-event knows how the input source is called, so it could generate a fairly nice error message.

**Comment by martinpitt** [comment]
Is this still good to merge, in the light of PR #4242?

**Comment by keszybz** [comment]
Yep, that's what they call "an unrelated fix".

**Comment by mbiebl** [comment]
@fbuihuu I tried backporting that patch for v215 (which we ship in Debian jessie, our current stable release), see https://github.com/mbiebl/systemd/commits/v215-CVE-2016-7796
Unfortunately the issue is still there. I.e. the systemd notification system still get's confused when sending an empty message and systemd eventually killing services like logind or journald.

Does this need a different fix for v215? Would appreciate any help with this.

**Comment by evverx** [comment]
@mbiebl , try

``` diff
diff --git a/src/core/manager.c b/src/core/manager.c
index 0cb2044..5a830bb 100644
--- a/src/core/manager.c
+++ b/src/core/manager.c
@@ -1394,13 +1394,15 @@ static int manager_dispatch_notify_fd(sd_event_source *source, int fd, uint32_t

                 n = recvmsg(m->notify_fd, &msghdr, MSG_DONTWAIT);
                 if (n <= 0) {
-                        if (n == 0)
-                                return -EIO;
+                        if (n == 0) {
+                                log_debug("Got zero-length notification message. Ignoring.");
+                                return 0;
+                        }

                         if (errno == EAGAIN || errno == EINTR)
                                 break;

-                        return -errno;
+                        return 0;
                 }

                 if (msghdr.msg_controllen < CMSG_LEN(sizeof(struct ucred)) ||
```

**Comment by fbuihuu** [comment]
@mbiebl v215 doesn't seem different from v210 in this regards.

We actually backported the fix for v210 (see https://github.com/openSUSE/systemd/commit/458d7ef2f53384ae4d32eb21bea8b9f58616e95a) and AFAICS the fix works as expected.

So I'm not sure why it wouldn't in your case....

**Comment by mbiebl** [comment]
@fbuihuu your (openSUSE) patch seems to work fine. Thanks for sharing.
I didn't get any watchdog timeouts anymore. Will apply that to the Debian package as well.

**Comment by fbuihuu** [comment]
@mbiebl well your version should have worked too AFAICS, I'm not sure to understand why this one works better....

**Comment by mbiebl** [comment]
@fbuihuu yeah, not sure why my backport didn't work. But since your's did, I didn't bother to investigate further.

**Comment by evverx** [comment]
@fbuihuu 

> For example pid1 won't be able to receive the WATCHDOG messages anymore and
> will kill all services supposed to send such messages.

`systemctl daemon-reexec` reenables all disabled event sources. No?

> If manager_dispatch_notify_fd() fails and returns an error then the handling of
> service notifications will be disabled entirely leading to a compromised system

What about `manager_dispatch_signal_fd`?

I'm not sure. Do we really need to ignore all `recvmsg`-errors (`n < 0`)?

**Comment by mbiebl** [comment]
@evverx In my case (i.e. v215 on Debian jessie) systemd remained in a broken state even after a daemon-reexec. A systemctl restart rsyslog for example, did hang and time out eventually. systemd-logind didn't properly recover either.

**Comment by evverx** [comment]
@mbiebl , I think the socket queue was full.
i.e. you can reproduce:

``` sh
root# export NOTIFY_SOCKET=/run/systemd/notify
root# cat /proc/sys/net/unix/max_dgram_qlen
512
root# for i in {1..514}; do echo $i; systemd-notify ""; done
1
2
...
447
...hangs...
^C

# install new version
root# systemctl daemon-reexec # into new version
root# systemd-notify ""
...hangs...
...the queue still full...
...we don't cleanup our queue on reexec...
```

**Comment by mbiebl** [comment]
@evverx that's possible. I did run systemd-notify in a loop for a while.

**Comment by evverx** [comment]
@mbiebl , oh, sorry, actually, I installed the new version into `/usr/local/lib/systemd/systemd` :) So, my reproducer doesn't make sense. Sorry

**Comment by fbuihuu** [comment]
@evverx 

> systemctl daemon-reexec reenables all disabled event sources. No?

I think so but not sure to get your point.

> What about manager_dispatch_signal_fd?

I think it needs the same fix, good point.

> I'm not sure. Do we really need to ignore all recvmsg-errors (n < 0)?

Which errors would need a different treatment ?

And in that case what would be the action that needs to be taken since disabling the handler is not an option ?

**Comment by evverx** [comment]
@fbuihuu 

> systemctl daemon-reexec reenables all disabled event sources. No?
> 
> > I think so but not sure to get your point.

Sorry, I mean, we can always run `daemon-reexec`. So

> the handling of
> service notifications will be disabled entirely leading to a compromised system.

is not so bad :-)

Well, I think I was wrong. Sorry for the noise.

But shouldn't we log something like

``` c
log_warning("Failed to receive notification message: %m, ignoring")
```

instead of the

``` c
log_error("Failed to receive notification message: %m")
```

?

**Comment by fbuihuu** [comment]
@evverx 

> But shouldn't we log something like
> `log_warning("Failed to receive notification message: %m, ignoring")`

Makes sense, indeed.

**Comment by poettering** [comment]
So, what precisely is this bug actually fixing? Did this ever become a problem IRL?

The reason why sd-event disables event sources on failure is to ensure we don't busy loop due to unprocessed IO events. I am pretty sure if the choice is between disabling an IO source or entering a busy loop, then disabling is the better default.

Now, I am pretty sure this should also be the right default for notify msg handling. Unless of course there were real-life cases where we got woken up, failed to process the message and then stopped processing anything else...

@fbuihuu can you elaborate on what made you make this change?

**Comment by keszybz** [comment]
In principle log_oom() could happen because of temporary resource exhaustion. It seems much better to only log a warning, since returning an error hoses the system.

For recv, yeah, we should not get an error. But if we get an error for any reason, propagating the error would also hose the system, so it seems better to try to continue.

**Comment by fbuihuu** [comment]
@poettering this was done by assuming that we can't realistically get a permanent error from recvmsg (which uses a Unix Datagram socket) and therefore any hypothetic errors are considered temporary by default. 

If this assumption is not true and recvmsg can return a permanent error in our case (which one ?) then I think it's still better to make this case special and still let try to recover for others (unlikely) errors because returning an error from the handler _will_ screw the system.

Regarding the oom error, see @keszybz answer: it's considered as a temporary failure.

**Comment by poettering** [comment]
> In principle log_oom() could happen because of temporary resource exhaustion. It seems much better to only log a warning, since returning an error hoses the system.

The log_oom() change is fine, as we know at that point that we have processed a datagram.

What I am just concerned about is that we busy loop when for some reason we can't read the incoming datagram (let's say because of some LSM or so sending us EPERM because of an fd attached to the datagram, or something like that). I am pretty sure it's still better to fail with an error in this case, than to just continue wanting to process the message over and over again...

I thought about that for quite some time when wondering what to do when sd-event IO handlers return an error. And I am pretty sure what holds back then still holds now: IO handlers should rather be turned off if an unprocessable event is seen rather than busy looping.

**Comment by poettering** [comment]
> @poettering this was done by assuming that we can't realistically get a permanent error from recvmsg (which uses a Unix Datagram socket) and therefore any hypothetic errors are considered temporary by default.

Well, we could actually get one, some internal kernel memory allocation issue, or a security check that's a bit too tight by some LSM or so, or anything like that.

I think the logic we should follow is this: if we get an error back from recvmsg() that indicates the epoll() wakeup was spurious (i.e. EAGAIN/EINTR), let's ignore things. If we get any other recvmsg() error, fail the event, as in this case we should assume that the datagram couldn't be dequeued, and we'd enter a busy loop. if an error happens after the recvmsg(), then log about it but do not fail the event, as in this case the busy loop chance is gone.

**Comment by poettering** [comment]
I posted #4304 now that alters the error handling so that non-spurious errors on recvmsg will be propagated, but spurious ones and errors after recvmsg are not. I hope this makes sense. Please have a look.

(That PR also adds two additional checks to the code, just to be safe)

**Pull request merged**